Add support for MCP23017 and redesign of tests

GoranLundberg opened the pull request:
- Add support for the MCP23017 16-bit (2x8) i2c expander.
   Supports both banks A and B. Should be able to drive two displays on one expander. 
   Don't have two displays without a backpack so I haven't tested it. Works fine with one.

- Redesign of tests.
  Using modules to easy add new tests and displaysizes.
  New style for command line options and parser.
  
  Example: `lcdtest.py i2c show_charmap expander=MCP23017 addr=0x20 cols=20 rows=4 gpio_bank=A`

  New files:
    - ledtest.py                             - main test file
    - ledtests/show_charmap.py  - module
    - ledtests/format_test_16x2    - module
    - ledtests/format_test_20x4    - modules
  
Old tests should be removed.

**Note: This needs some fine tuning and testing!
          Would appreciate comments on improvements on code and idea/concept.
          You might want to rename some files/variables etc. Haven't tested it on PCF8547**

GoranLundberg commented:
Ooops. Got some old changes as well by accident. Sorry!

GoranLundberg commented:
Added:
- fix errors in custom characters (CGRAM) in the A02 codec.

dbrgn commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+    elif test == 'format_test':
+        if ((cols == 20) and (rows == 4)):
+            format_test_20x4.run(lcd)
+        elif ((cols == 16) and (rows == 2)):
```
You could drop those parentheses here and on line 161:

`if cols == 16 and rows == 2`

dbrgn commented on lcdtests/show_charmap.py:
```diff
@@ -0,0 +1,82 @@
+    safe_input = input
+
+
+from RPLCD import i2c, gpio
```
I think this import is not necessary anymore, right? I think it can also be removed from the format test files.

GoranLundberg replied:
Yes, will remove it.

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+
+        print('<options> gpio options:')
+        print('')
+        print('   mode     - GPIO numbering mode, either GPIO.BOARD or GPIO.BCM, default=GPIO.BCM')
```
Okey. Will do!

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+        print('   data    - Data (d0-d7) gpio pin numbers (4 or 8 numbers depending')
+        print('                    on if you want 4 or 8 bit mode, separated by comma)')
+        print('                    Example: data=1,2,3,4,5,6,7,8 (for 8-bit mode)')
+        print('                             data=4,5,6,7,8 (for 4-bit mode)')
```
Oooops!

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+        print('              characters, it probably uses the A00 charmap, otherwise A02.')
+        print('              default=A00')
+        print('')
+        print('Options format: name=value e.g. expander=MCP23008 (Example below)')
```
Will do.

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+    # Parse options into a dictionary
+    options = {}
+    if sys.version_info > (3, 0):
+            options = dict([arg.split(sep='=', maxsplit=1) for arg in sys.argv[3:]])
```
In case the 2.x code gets depreciated in future versions. Can remove it.

dbrgn replied:
Ah, I see. I think that's something that automated tools can catch, once we can drop Python 2 support.

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+        # i2c options
+        i2c_expander = options.pop('expander')
+        address = int(options.pop('addr'), 16)
+        port = int(options.pop('port', '1'))
```
Ooops!

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+        port = int(options.pop('port', '1'))
+
+        lcd = i2c.CharLCD(i2c_expander, address, cols=cols, port=port,
+                          rows=rows, expander_params=options)
```
Ooops!

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+        pins_data = {}
+        if sys.version_info > (3, 0):
+            pins_data = data.split(sep=',')
+            print(pins_data)
```
Yup, already removed in my local code.

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+            pins_data = data.split(',')
+        # Convert data pins to int
+        for i in range(len(pins_data)):
+            pins_data[i] = int(str(pins_data[i]))
```
Ah, didn't know you could do that in Python!

GoranLundberg commented on lcdtest.py:
```diff
@@ -0,0 +1,168 @@
+        lcd = gpio.CharLCD(pin_rs=rs, pin_rw=rw, pin_e=e, pins_data=pins_data,
+                       pin_backlight=bl, numbering_mode=mode, cols=cols, rows=rows,
+                       charmap=charmap)
+        pass
```
hehe, yeah :)

GoranLundberg commented:
- Fixed issues you mentioned
- Added more error checking in lcdtest.py
- Removed old tests

- Removed default gpio pins and numbering_mode in gpio.py

dbrgn commented on lcdtest.py:
```diff
@@ -102,67 +137,77 @@ def print_usage():
+            testsuite_16x2.run(lcd)
         else:
-            print_usage()
+            print_usage(str(cols) + 'x' + str(rows) + ' displays not supported in this test.')
```
Again, as noted above, it's easier to use string formatting:

```python
print_usage('%dx%d displays not supported in this test.' % (cols, rows))
```

There are some other instances of concatenation in this part of the code that could also be simplified like that.

dbrgn commented on lcdtest.py:
```diff
@@ -0,0 +1,204 @@
+        rs = int(options_pop('rs'))
+        e = int(options_pop('e'))
+        rw = options_pop('rw', 'None')
+        rw = None if rw is 'None' else int(rw)
```
Another case where `rw is 'None'` should be replaced with `rw == 'None'`.

dbrgn commented on lcdtests/show_charmap.py:
```diff
@@ -39,34 +37,10 @@
-    print('<cols>      The number of columns on your LCD, e.g. 16')
-    print('')
-    sys.exit(1)
+#from RPLCD import i2c, gpio
```
Dead code :skull_and_crossbones:

dbrgn commented on lcdtests/testsuite_20x4.py:
```diff
@@ -0,0 +1,165 @@
+"""
+from __future__ import print_function, division, absolute_import, unicode_literals
+
+#from RPLCD import gpio, i2c
```
Dead code :skull_and_crossbones:

dbrgn commented:
Merged in 3edfd3ddc3e12ffa8d1be5c069d1b9622d4b6adc :)
